contrib/endpoints/src/api_manager/check_security_rules.cc

```diff
@@ -14,92 +14,271 @@
 //
 ////////////////////////////////////////////////////////////////////////////////
 #include "contrib/endpoints/src/api_manager/check_security_rules.h"
+#include "contrib/endpoints/src/api_manager/auth/lib/json_util.h"
+#include <sstream>
+#include <iostream>
 
-#include <string>
-
-#include "contrib/endpoints/include/api_manager/api_manager.h"
-#include "contrib/endpoints/include/api_manager/request.h"
-
+using ::google::api_manager::auth::GetProperty;
+using ::google::api_manager::auth::GetStringValue;
 using ::google::api_manager::utils::Status;
+using ::google::protobuf::util::error::Code;
 
 namespace google {
 namespace api_manager {
 
-namespace {
-
 const char kFirebaseServerStaging[] =
     "https://staging-firebaserules.sandbox.googleapis.com/";
 
-// An AuthzChecker object is created for every incoming request. It does
-// authorizaiton by calling Firebase Rules service.
-class AuthzChecker : public std::enable_shared_from_this<AuthzChecker> {
- public:
-  AuthzChecker(std::shared_ptr<context::RequestContext> context,
-               std::function<void(Status status)> continuation);
+const char kFirebaseService[] =
+    "/google.firebase.rules.v1.FirebaseRulesService";
+
+const char kFailedFirebaseReleaseFetch[] = "Failed to fetch Firebase Release";
+const char kFailedFirebaseTest[] = "Failed to execute Firebase Test";
+const char kInvalidResponse[] = "Invalid JSON response from Firebase Service";
+
+AuthzChecker::AuthzChecker(ApiManagerEnvInterface *env,
+                           auth::ServiceAccountToken *sa_token)
+    : env_(env),
+      sa_token_(sa_token) {}
+
+void AuthzChecker::Check(
+    std::shared_ptr<context::RequestContext> context,
+    std::function<void(Status status)> final_continuation) {
+  // TODO: Check service config to see if "useSecurityRules" is specified.
+  // If so, call Firebase Rules service TestRuleset API.
 
-  void Check();
+  if (!context->service_context()->RequireAuth() ||
+      context->method() == nullptr || !context->method()->auth()) {
+    env_->LogDebug(
+        std::string("Autherization and JWT validation was not performed")
+        + " skipping authorization");
+    final_continuation(Status::OK);
+    return;
+  }
 
- private:
-  // Helper function to send a http GET request.
-  void HttpFetch(const std::string &url, const std::string &request_body,
-                 std::function<void(Status, std::string &&)> continuation);
+  // Fetch the Release attributes.
+  auto pChecker = GetPtr();
```

Comment:
Hmm. This code is again similar to check_auth.cc below. Let  me know if you want me to fix it in both places.

void AuthChecker::DiscoverJwksUri(const std::string &url) {
  auto pChecker = GetPtr();
  HttpFetch(url, [pChecker](S